File: smtp/operation/workflow/smtp.js
/*
 *
 * INFIVERVE TECHNOLOGIES PTE LIMITED CONFIDENTIAL
 * _______________________________________________
 *
 *  (C) INFIVERVE TECHNOLOGIES PTE LIMITED, SINGAPORE
 *  All Rights Reserved.
 *  Product / Project: Flint IT Automation Platform
 *  NOTICE:  All information contained herein is, and remains
 *  the property of INFIVERVE TECHNOLOGIES PTE LIMITED.
 *  The intellectual and technical concepts contained
 *  herein are proprietary to INFIVERVE TECHNOLOGIES PTE LIMITED.
 *  Dissemination of this information or any form of reproduction of this material
 *  is strictly forbidden unless prior written permission is obtained
 *  from INFIVERVE TECHNOLOGIES PTE LIMITED, SINGAPORE.
*/

log.trace("Starting the execution of 'flint-util:smtp:operation:workflow:smtp.js'");

//Connector
connector_name = 'smtp';
connector_call = call.connector(connector_name);
log.info("Connector Name: " + connector_name);

//Action
action = "send";
log.info("Action: " + action);
connector_call.set("action", action);

input_clone = JSON.parse(input);

if (input_clone.hasOwnProperty("protocol_connection")) {

    protocol_connection = input_clone["protocol_connection"];
    encryptedCredentials = protocol_connection["encryptedCredentials"];
    //Username
    username = encryptedCredentials["username"];
    if (username != "" && username != null) {
        connector_call.set("username", username);
        log.info("Username: " + username);
    }
    else {
        log.error("Username is null or a empty string.");
    }

    //Password
    password = encryptedCredentials["password"];
    if (password != "" && password != null) {
        connector_call.set("password", password);
        log.info("Password is given.");
    }
    else {
        log.error("Password is an empty string or a null");
    }

    //Target
    target = encryptedCredentials["target"];
    if (target != "" && target != null) {
        connector_call.set("target", target);
        log.info("Target: " + target);
    }
    else {
        log.error("Hostname is an empty string or null.");
    }

    //Port
    port = encryptedCredentials["port"];
    if (port != null && port != "") {
        connector_call.set("port", port);
        log.info("Port: " + port);
    }
    else {
        log.error("Port is an empty string or null");
    }

    //From
    if (input_clone.hasOwnProperty("from")) {
        from = input.get("from");
        if (from != null && from != "") {
            connector_call.set("from", from);
            log.info("From: " + from);
        }
        else {
            log.error("From is null or an empty string");
        }
    }
    else {
        log.error("Input JSON doesn't contain 'from' key");
    }

    //To
    if (input_clone.hasOwnProperty("to")) {
        to = input.get("to");
        if (to != null && to != "") {
            connector_call.set("to", to);
            log.info("To: " + to);
        }
        else {
            log.error("To is null or an empty string");
        }
    }
    else {
        log.error("Input JSON doesn't contain 'to' key");
    }

    //Subject
    if (input_clone.hasOwnProperty("subject")) {
        subject = input.get("subject");
        if (subject != null && subject != "") {
            connector_call.set("subject", subject);
            log.info("Subject: " + subject);
        }
        else {
            log.error("Subject is null or an empty string");
        }
    }
    else {
        log.error("Input JSON doesn't contain 'subject' key");
    }

    //Body
    if (input_clone.hasOwnProperty("body")) {
        body = input.get("body");
        if (body != null && body != "") {
            connector_call.set("body", body);
            log.info("Body: " + body);
        }
        else {
            log.error("Body is null or an empty string");
        }
    }
    else {
        log.error("Input JSON doesn't contain 'body' key");
    }

    //cc
    if (input_clone.hasOwnProperty("cc")) {
        cc = input.get("cc");
        if (cc != null && cc != "") {
            index_cc = cc.indexOf(",");
            if (index_cc != -1) {
                cc = cc.split(",");
                connector_call.set("cc", cc);
                log.info("cc: " + cc);
            }
            else {
                cc_array = [];
                cc_array.push(cc);
                connector_call.set("cc", cc_array);
                log.info("cc: " + cc_array);
            }
        }
        else {
            log.info("cc is null or an empty string");
        }
    }
    else {
        log.info("Input JSON doesn't contain 'cc' key");
    }

    //bcc
    if (input_clone.hasOwnProperty("bcc")) {
        bcc = input.get("bcc");
        if (bcc != null && bcc != "") {
            index_bcc = bcc.indexOf(",");
            if (index_bcc != -1) {
                bcc = bcc.split(",");
                connector_call.set("bcc", bcc);
                log.info("bcc: " + bcc);
            }
            else {
                bcc_array = [];
                bcc_array.push(bcc);
                connector_call.set("bcc", bcc_array);
                log.info("bcc: " + bcc_array);
            }
        }
        else {
            log.info("bcc is null or an empty string");
        }
    }
    else {
        log.info("Input JSON doesn't contain 'bcc' key");
    }

    //Attachments
    if (input_clone.hasOwnProperty("attachments")) {
        attachments = input.get("attachments");
        log.trace("Current attachment value " + typeof attachments)
        if (attachments != null && attachments != "" && attachments != []) {
            index_attachment = attachments.indexOf(",");
            if (index_attachment != -1) {
                attachments = attachments.split(",");
                connector_call.set("attachments", attachments);
                log.info("Attachments: " + attachments);
            }
            else {
                attachments_array = [];
                attachments_array.push(attachments);
                connector_call.set("attachments", attachments_array);
                log.info("Attachments: " + attachments_array);
            }
        }
        else {
            log.info("attachments is null or an empty string");
        }
    }
    else {
        log.info("Input JSON doesn't contain 'attachments' key");
    }

    //Connector call
    response = connector_call.sync();

    //Response Meta Parameters
    response_exitcode = response.exitcode();
    response_message = response.message();

    //Response Result
    result = response.get("result");

    if (response_exitcode == 0) {
        log.info("Email sent Successfully!");
        user_message = "email sent successfully.";
        output.set("result", result).set("user_message", user_message).set("exit-code", response_exitcode);
        log.trace("Finished executing 'flint-util:smtp:operation:workflow:smtp.js' successfully");
    }
    else {
        log.error("Failure in sending email");
        output.set("error", response_message).set("exit-code", response_exitcode);
        log.trace("Finished executing 'flint-util:smtp:operation:workflow:smtp.js' with errors");
    }
}
else {
    log.error("Protocol Connection is not given.");
}

File: smtp/validation/smtp.rb
=begin
##########################################################################
#
#  INFIVERVE TECHNOLOGIES PTE LIMITED CONFIDENTIAL
#  __________________
# 
#  (C) INFIVERVE TECHNOLOGIES PTE LIMITED, SINGAPORE
#  All Rights Reserved.
#  Product / Project: Flint IT Automation Platform
#  NOTICE:  All information contained herein is, and remains
#  the property of INFIVERVE TECHNOLOGIES PTE LIMITED.
#  The intellectual and technical concepts contained
#  herein are proprietary to INFIVERVE TECHNOLOGIES PTE LIMITED.
#  Dissemination of this information or any form of reproduction of this material
#  is strictly forbidden unless prior written permission is obtained
#  from INFIVERVE TECHNOLOGIES PTE LIMITED, SINGAPORE.
=end

# begin
@log.info('Started execution of SMTP script !!')
begin
    if @input.type == 'application/xml'
        @connector_name = @input.get('/connector_name/text()') # Name of the smtp connector
        @from = @input.get('/from/text()')                     # Address of mail sender
        @to = @input.get('/to/text()')                         # Address of mail reciever
        @subject = @input.get('/subject/text()')               # Subject of the mail
        @body = @input.get('/body/text()')                     # Body of the mail
        @cc = @input.get('/cc/text()')                         # Carbon copy send to address other than main address
        @bcc = @input.get('/bcc/text()')                       # Blank carbon copy send to address other than main address
        @attachments = @input.get('/attachments/text()')       # Files to be added during mail
    else
        @connector_name = @input.get('connector_name')         # Name of the smtp connector
        @from = @input.get('from')                             # Address of mail sender
        @to = @input.get('to')                                 # Address of mail reciever
        @subject = @input.get('subject')                       # Subject of the mail
        @body = @input.get('body')                             # Body of the mail
        @cc = @input.get('cc')                                 # Carbon copy send to address other than main address
        @bcc = @input.get('bcc')                               # Blank carbon copy send to address other than main address
        @attachments = @input.get('attachments')               # Files to be added during mail
    end

    @connector_name = 'email' if @connector_name.nil? || @connector_name.empty?
    @from = 'email address' if @from.nil? || @from.empty?
    @to = 'email address' if @to.nil? || @to.empty?
    @subject = 'This is subject' if @subject.nil? || @subject.empty?
    @body = 'This is body' if @body.nil? || @body.empty?
    @cc = 'email address' if @cc.nil? || @cc.empty?
    @bcc = 'email address' if @bcc.nil? || @bcc.empty?
    if @attachments.nil? || @attachments.empty?
        @attachments = '/home/kamaljeet/Desktop/add.rb'
    end

    @log.info("Flintbit input parameters are, connector_name :: #{@connector_name}|
    from :: #{@from}| to   :: #{@to}| subject :: #{@subject}|
    body :: #{@body}| cc :: #{@cc}| bcc :: #{@bcc}| attachments :: #{@attachments}")
    @log.trace('Calling SMTP Connector...')

    response = @call.connector(@connector_name)
                    .set('cc', @cc)
                    .set('bcc', @bcc)
                    .set('subject', @subject)
                    .set('from', @from)
                    .set('to', @to)
                    .set('body', @body)
                    .set('action', 'send')
                    .set('attachments', @attachments).sync

    # SMTP Connector Response Meta Parameters
    response_exitcode = response.exitcode    # Exit status code
    response_message = response.message      # Execution status message

    # SMTP Connector Response Parameters
    result = response.get('result') # Response Body

    if response.exitcode == 0
        @log.info("Success in executing SMTP Connector where, exitcode :: #{response_exitcode} | message ::  #{response_message}")
        @log.info("SMTP Response Body ::#{response_message}")
        @output.set('result', response_message)
    else
        @log.error("Failure in executing SMTP Connector where, exitcode :: #{response_exitcode} | message ::  #{response_message}")
        @log.error('Failed')
        @output.set('result', response_message)
    end
rescue Exception => e
    @log.error(e.message)
    @output.set('exit-code', 1).set('message', e.message)
end
@log.info('Finished execution of SMTP script !!')
# end
